tools/build.py

```diff
@@ -0,0 +1,681 @@
+                    str(test_dir / "extensions_test.exe"),
+                    str(test_dir / "ocos_test.exe")
+                ]
+
+                _run_subprocess(
+                    [
+                        "vstest.console.exe",  # must be in the PATH. run this script from a VS dev shell.
+                        "--parallel",
+                        f"--TestAdapterPath:{str(adapter)}",
+                        "/Logger:trx",
+                        "/Enablecodecoverage",
+                        "/Platform:x64",
+                        f"/Settings:{str(REPO_DIR / 'test' / 'codeconv.runsettings')}",
+                    ]
+                    + executables,
+                    cwd=build_dir,
+                )
+
+            else:
+                executables = ["extensions_test", "ocos_test"]
+                for exe in executables:
+                    _run_subprocess([str(cwd / exe)], cwd=cwd)
+
+
+def main():
+    log.debug("Command line arguments:\n  {}".format(" ".join(shlex.quote(arg) for arg in sys.argv[1:])))
+
+    args = _parse_arguments()
+    cmake_extra_defines = _flatten_arg_list(args.cmake_extra_defines)
+    cross_compiling = args.arm or args.arm64 or args.arm64ec or args.android
+
+    # If there was no explicit argument saying what to do, default
+    # to update, build and test (for native builds).
+    if not (args.update or args.clean or args.build or args.test):
+        log.debug("Defaulting to running update, build [and test for native builds].")
+        args.update = True
+        args.build = True
+        if cross_compiling:
+            args.test = args.android_abi == "x86_64" or args.android_abi == "arm64-v8a"
+        else:
+            args.test = True
+
+    if args.skip_tests:
+        args.test = False
+
+    if args.build_wasm_static_lib:
+        args.build_wasm = True
+
+    if args.build_wasm:
+        if not args.disable_wasm_exception_catching and args.disable_exceptions:
+            # When '--disable_exceptions' is set, we set '--disable_wasm_exception_catching' as well
+            args.disable_wasm_exception_catching = True
+        if args.test and args.disable_wasm_exception_catching and not args.minimal_build:
+            raise UsageError("WebAssembly tests need exception catching enabled to run if it's not minimal build")
+
+    if args.android and is_windows():
+        if args.cmake_generator != "Ninja":
+            log.info("Setting cmake_generator to Ninja, which is required when cross-compiling Android on Windows.")
+            args.cmake_generator = "Ninja"
+
+    configs = set(args.config)
+
+    # setup paths and directories
+    # cmake_path can be None. For example, if a person only wants to run the tests, they don't need cmake.
+    cmake_path = _resolve_executable_path(args.cmake_path)
+    build_dir = args.build_dir
+
+    if args.update or args.build:
+        for config in configs:
+            os.makedirs(_get_build_config_dir(build_dir, config), exist_ok=True)
```

Comment:
also done in _generate_build_tree, probably not needed here

Reply:
this one runs first so removed the one in _generate_build_tree